python/ray/experimental/sgd/param_server.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import logging
+
+import numpy as np
+
+import ray
+from ray.experimental.sgd.util import Timeline, fetch, warmup
+
+logger = logging.getLogger(__name__)
+
+
+@ray.remote(num_cpus=0)
+class ParameterServer(object):
+    """Helper class for ray.experimental.sgd.DistributedSGD."""
+
+    def __init__(self, num_workers, tid):
+        self.num_sgd_workers = num_workers
+        self.acc_counter = 0
+        self.timeline = Timeline(tid)
+        # TODO(ekl) get this to work again so we get ray events
+        # self.timeline.patch_ray()
+
+    def initialize(self, shard_shape):
+        """Resets the gradient buffer to zeros."""
+        self.accumulated = np.zeros(shard_shape, dtype=np.float32)
+
+    def prefetch(self, oids):
+        """Tell plasma to prefetch the given object ids over the network."""
+        self.timeline.reset()
+        self.timeline.start("prefetch")
+        fetch(oids)
+        self.timeline.end("prefetch")
+
+    def add_spinwait(self, grad_shard_ids):
+        """Optimized version of add() that operates on multiple grads."""
+        self.timeline.start("add_spinwait")
+        plasma_ids = [ray.pyarrow.plasma.ObjectID(x) for x in grad_shard_ids]
+        while plasma_ids:
+            for p in plasma_ids:
+                if ray.worker.global_worker.plasma_client.contains(p):
+                    self.timeline.start("get_buffers")
+                    grads = ray.worker.global_worker.plasma_client.get(p)
+                    self.accumulated += grads
+                    self.acc_counter += 1
+                    self.timeline.end("get_buffers")
+                    plasma_ids.remove(p)
+                    break
+        self.timeline.end("add_spinwait")
+
+    def add(self, grad_shard_id):
+        """Add the given gradient value to the accumulated gradients."""
+        self.timeline.start("add")
+        self.timeline.start("get_buffers")
+        oid = ray.pyarrow.plasma.ObjectID(grad_shard_id)
+        grads = ray.worker.global_worker.plasma_client.get(oid)
+        self.timeline.end("get_buffers")
+        self.accumulated += grads
+        self.acc_counter += 1
+        self.timeline.end("add")
+
+    def get(self, object_id):
+        """Put the accumulated gradients to the given object id."""
+        self.timeline.start("get")
```

Comment:
I think this kind of start and end code can be replaced by a decorator. It is also good for later refactoring.

Reply:
Yeah, let's not fix this code though. Ultimately it should just be removed and replaced with the `ray.profile` API.